packages/flutter_tools/lib/src/base/build.dart

```diff
@@ -0,0 +1,136 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert' show JSON;
+
+import 'package:crypto/crypto.dart' show md5;
+import 'package:meta/meta.dart' show required;
+
+import 'file_system.dart';
+
+/// A function that writes a set of outputs, given a set of input files.
+typedef Future<Null> BuildFunction(Set<File> inputs);
+
+class BuildException implements Exception {
+  BuildException(this.message, {this.cause}) {
+    assert(message != null);
+  }
+
+  final String message;
+  final dynamic cause;
+
+  @override
+  String toString() => message;
+}
+
+/// A build action that generates output artifacts given a set of inputs and a
+/// build function.
+///
+/// Build actions enforce the following rules:
+/// 1. There must be at least one input.
+/// 2. There must be at least one output.
+/// 3. No input may also be an output.
+/// 3. All declared inputs exist prior to the build.
+/// 4. All declared outputs exist after the build.
+///
+/// If a [checksumPath] is specified, a checksum is computed for every input
+/// file prior to the build, and recorded to the path specified. If all output
+/// files are present and the input file checksums are unchanged from the
+/// previous build, the build action is skipped, and the existing outputs are
+/// assumed valid.
+///
+/// As a consequence, if [checksumPath] is specified, all output files must be
+/// purely a function of the declared input files; otherwise, stale build
+/// results may be returned.
+class BuildAction {
+  BuildAction({
+    @required this.inputPaths,
+    @required this.outputPaths,
+    @required this.buildFunction,
+    String checksumPath,
+  }) : _checksumPath = checksumPath {
+    assert(inputPaths != null && inputPaths.isNotEmpty);
+    assert(outputPaths != null && outputPaths.isNotEmpty);
+    assert(buildFunction != null);
+  }
+
+  final Set<String> inputPaths;
+  final Set<String> outputPaths;
+  final BuildFunction buildFunction;
+  final String _checksumPath;
+
+  Future<Null> build() async {
+    await _validate();
+    try {
+      final List<File> inputs = inputPaths.map(fs.file).toList(growable: false);
+      final List<File> outputs = outputPaths.map(fs.file).toList(
+          growable: false);
+
+      await _assertFilesExist(inputs, fileType: 'input');
+
+      final Map<String, int> oldChecksums = await _readChecksums(_checksumPath);
+      final Map<String, int> newChecksums = await _computeChecksums(inputs);
+      if (!_checksumsMatch(oldChecksums, newChecksums) ||
+          (await _missingFiles(outputs)).isNotEmpty) {
+        await buildFunction(inputs.toSet());
+        _writeChecksums(_checksumPath, newChecksums);
+      }
+      await _assertFilesExist(outputs, fileType: 'output');
+    } catch (e, s) {
+      throw new BuildException('ERROR: $e\n$s');
+    }
+  }
+
+  Future<Null> _validate() async {
```

Comment:
Good point. Initially I had existence checks for the files in here, but since this is now purely in-memory and any i/o is limited to the build() method, makes sense. Done.